Add pulseaudio to a stubdomain

easydozen opened the pull request:
QubesOS/qubes-issues#2624

easydozen committed:
add pulseaudio tarball and BR to the rpm spec template

Signed-off-by: Dmitry Fedorov <d.fedorov@tabit.pro>

easydozen committed:
add qubes-gui-agent-linux submodule to get module-vchan-sink

Signed-off-by: Dmitry Fedorov <d.fedorov@tabit.pro>

easydozen committed:
replace pulseaudio Makefile vars to provide minimalistic build
with module-vchan-sink

Signed-off-by: Dmitry Fedorov <d.fedorov@tabit.pro>

easydozen committed:
conditionally use qubesdb dependency in vchan-sink module

Signed-off-by: Dmitry Fedorov <d.fedorov@tabit.pro>

easydozen committed:
extract, patch, configure and build PA in appropriate for qubes-builder manner

Signed-off-by: Dmitry Fedorov <d.fedorov@tabit.pro>

easydozen committed:
build qemu with our own PA headers and libraries
no need to use external PA BR

Signed-off-by: Dmitry Fedorov <d.fedorov@tabit.pro>

easydozen committed:
install PA to the stubdom rootfs

Signed-off-by: Dmitry Fedorov <d.fedorov@tabit.pro>

easydozen committed:
implement intel-hda soundhw names that exist in libvirt

Signed-off-by: Dmitry Fedorov <d.fedorov@tabit.pro>

easydozen committed:
run PA if domain args contains soundhw

Signed-off-by: Dmitry Fedorov <d.fedorov@tabit.pro>

marmarek commented on pulseaudio/patches/pulseaudio-add-qubesdb-client-ifdef.patch:
```diff
@@ -0,0 +1,28 @@
+--- a/src/modules/module-vchan-sink.c	2020-05-27 18:06:41.683862501 +0300
```
Please apply this to https://github.com/QubesOS/qubes-gui-agent-linux/ directly, and add relevant `-D` flag to the makefile there. It will be much easier to maintain this way, instead of refreshing this patch each time module-vchan-sink.c changes.

marmarek commented on qemu/patches/qemu-register-ich6-ich9-soundhw.patch:
```diff
@@ -0,0 +1,55 @@
+     type_register_static(&intel_hda_info_ich6);
+     type_register_static(&intel_hda_info_ich9);
+     type_register_static(&hda_codec_device_type_info);
+-    pci_register_soundhw("hda", "Intel HD Audio", intel_hda_and_codec_init);
```
I'm confused by this patch. This rather looks like a libvirt bug than qemu issue. I think the proper fix should be extending libxl libvirt driver with similar function as qemu libvirt driver has here:
https://github.com/libvirt/libvirt/blob/6f60d2a8503ce8c624bce6b53bf7b68476f5056f/src/qemu/qemu_command.c#L4019-L4056

But, that's probably for the time when libxl will migrate to `-device` option for sound card (names seems to be different). For now, I think it's ok to patch it this way, but please do not break original device name (do not remove the line above).

marmarek commented on Makefile.vars:
```diff
@@ -1,3 +1,4 @@
 QEMU_VERSION := 4.2.0
 LINUX_VERSION := 5.4.30
 BUSYBOX_VERSION := 1.31.1
+PULSEAUDIO_VERSION := 13.99.1
```
You also need to add pulseaudio to `URLS` in the `Makefile`. See other files there - url should point at a signature file (if exists, which doesn't look to be the case) or alternatively add a `.sha512` file with a hash.

BTW this should be detected by travis, as the build would fail now. But for some reason travis didn't built it at all. Looking at a log in their web interface I see your build marked as "Abuse detected"!?

easydozen commented:
> BTW this should be detected by travis, as the build would fail now. But for some reason travis didn't built it at all. Looking at a log in their web interface I see your build marked as "Abuse detected"!?

Thanks for mentioning. My previous opensource activity was mainly focused around another platform (abf.io/din) with another rules. I don't know much about github tools. Travis support didn't reveal the cause. Maybe I was flagged on [this PR](https://github.com/QubesOS/qubes-core-admin-client/pull/140). I forgot to sign my PR, then chose wrong key, did all the mess with force push. Anyway, sorry for that.

easydozen committed:
move vchan-sink patch to qubes-gui-agent-linux

easydozen committed:
add sha512 checksum for pulseaudio

easydozen committed:
do not replace original hda soundhw

easydozen committed:
add pulseaudio URL

marmarek commented on Makefile:
```diff
@@ -14,7 +14,8 @@ FETCH_CMD := wget --no-use-server-timestamps -q -O
     https://kernel.org/pub/linux/kernel/v$(firstword $(subst ., ,$(LINUX_VERSION))).x/linux-$(LINUX_VERSION).tar.sign \
-    https://busybox.net/downloads/busybox-$(BUSYBOX_VERSION).tar.bz2.sig
+    https://busybox.net/downloads/busybox-$(BUSYBOX_VERSION).tar.bz2.sig \
+    http://freedesktop.org/software/pulseaudio/releases/pulseaudio-$(PULSEAUDIO_VERSION).tar.xz
```
Please use https url.

marmarek reviewed (changes requested):
Since I've merged https://github.com/QubesOS/qubes-gui-agent-linux/pull/98 already, please update submodule here to let travis build the right version.

easydozen committed:
update qubes-gui-agent-linux submodule

easydozen committed:
use https URL

marmarek committed:
Avoid copying libpulse* from the host system

Set LD_LIBRARY_PATH pointing at local pulseaudio built libraries, to
avoid dracut-install copying them from the host system when available.
This has an extra effect of having libpulse installed by dracut-install
- no need for manual copy anymore.
Since dracut-install preserves original path, remove leftover /home from
within initramfs root dir.

easydozen commented:
Pulseaudio modules (module-native-protocol-unix, module-vchan-sink) not available in stubdomain since the last commit. They are not collected as dependencies to Pulseaudio binary.

marmarek commented:
Ah, you're right. While fixing one bug, I've introduced another. Let me re-add installing modules.

What would be helpful is to add automatic tests for this whole thing. There are already some for pulseaudio running inside target VM directly (https://github.com/QubesOS/qubes-core-admin/blob/master/qubes/tests/integ/vm_qrexec_gui.py#L411-L584). Perhaps it wouldn't be hard to do similar thing for pulseaudio in stubdomain? Something like enabling this feature and then starting pulseaudio within the VM to connect to emulated audio card instead of module-vchan-sink? Would you like to try?

easydozen commented:
> Would you like to try?

Surely. I'll do my best.

marmarek commented:
Here is some extra docs how to run them: https://www.qubes-os.org/doc/automated-tests/#qubes-40 (see also section above about running individual tests)

marmarek committed:
Restore installing pulseaudio modules

Use dracut install for this too, to make sure all dependencies are
installed too.

easydozen committed:
Pulseaudio: add ich6 preset

There are a variety of audiodev parameters that can be used to achieve
good sound quality in qemu. Unfortunately at the moment there is no safe
libvirt/libxl method to pass them to a stubdom.
Default hardcoded parameters for chosen audio model suggested here.
